src/shell/eventlogger.c

```diff
@@ -0,0 +1,223 @@
+    if (++ev->refcount == 1 && ev->ops.busy)
+        (*ev->ops.busy) (ev, ev->arg);
+    return 0;
+}
+
+static void eventlog_batch_error (struct eventlog_batch *batch, int errnum)
+{
+    struct eventlogger *ev = batch->ev;
+    json_t *entry = zlist_first (batch->entries);
+    while (entry) {
+        if (ev->ops.err)
+            (*ev->ops.err) (ev, entry);
+        entry = zlist_next (batch->entries);
+    }
+
+    eventlogger_batch_complete (batch->ev, batch);
+}
+
+static void commit_cb (flux_future_t *f, void *arg)
+{
+    struct eventlog_batch *batch = arg;
+    if (flux_future_get (f, NULL) < 0) {
+        eventlog_batch_error (batch, errno);
+        return;
+    }
+    eventlogger_batch_complete (batch->ev, batch);
+    flux_future_destroy (f);
+}
+
+static void
+timer_cb (flux_reactor_t *r, flux_watcher_t *w, int revents, void *arg)
+{
+    struct eventlog_batch *batch = arg;
+    flux_t *h = batch->ev->h;
+    flux_future_t *f = NULL;
+
+    if (!(f = flux_kvs_commit (h, NULL, 0, batch->txn))
+        || flux_future_then (f, -1., commit_cb, batch) < 0) {
+        eventlog_batch_error (batch, errno);
+        return;
+    }
+    batch->ev->current = NULL;
+}
+
+static struct eventlog_batch * eventlog_batch_create (struct eventlogger *ev)
+{
+    struct eventlog_batch *batch = calloc (1, sizeof (*batch));
+    if (!batch)
+        return NULL;
+    flux_reactor_t *r = flux_get_reactor (ev->h);
+    batch->ev = ev;
+    batch->entries = zlist_new ();
+    batch->txn = flux_kvs_txn_create ();
+    batch->timer = flux_timer_watcher_create (r,
+                                              ev->timeout, 0.,
+                                              timer_cb,
+                                              batch);
+    if (!batch->entries || !batch->txn || !batch->timer) {
+        eventlog_batch_destroy (batch);
+        return NULL;
+    }
+    flux_watcher_start (batch->timer);
+    return batch;
+}
+
+void eventlogger_destroy (struct eventlogger *ev)
+{
+    if (ev) {
+        if (ev->pending)
+            zlist_destroy (&ev->pending);
+        free (ev);
+    }
+}
+
+struct eventlogger *eventlogger_create (flux_t *h,
+                                        double timeout,
+                                        struct eventlogger_ops *ops,
+                                        void *arg)
+{
+    struct eventlogger *ev = calloc (1, sizeof (*ev));
+    if (ev) {
+        ev->pending = zlist_new ();
+        if (!ev->pending) {
+            eventlogger_destroy (ev);
+            return NULL;
+        }
+        ev->h = h;
+        ev->timeout = timeout;
+        ev->current = NULL;
+        ev->ops = *ops;
+        ev->arg = arg;
+    }
+    return ev;
+}
+
+int eventlogger_append (struct eventlogger *ev,
+                        int flags,
+                        const char *path,
+                        const char *name,
+                        const char *context)
+{
+    int rc = -1;
```

Comment:
i think the flow of this function is a little weird with the constant checks for `async`.  Would it just be better to have a different function call for each case, accepting the small bit of duplicate code?

Reply:
I was working on that (sorry, some of the actual code here isn't quite ready for review as I called out in the description). I'll have to see what works best. The problem with duplicating code is duplicating bugs.